### PR TITLE
sys/evtimer: introduce ZTIMER_MSEC as timer backend

### PR DESCRIPTION
**Pull request by jue89**
### Contribution description

Inspired by the implementation of `xtimer_on_ztimer`, I tried the same thing for `evtimer`.

According to evtimer's description, it has been built to allow for long running timers. Thus, I made an adaption to the ZTIMER_MSEC clock which should perfectly meet this requirement.

This PR is still WIP ... not all tests are green.

### Testing procedure

Run `tests/evtimer_msg` with the following command line:
`USEMODULE="ztimer_usec evtimer_on_ztimer" make all test`

(`tests/evtimer_underflow` doesn't work on `native`.)

### Issues/PRs references

~~Requires #13666~~

### CODE REVIEW

**Comment by jue89** [comment]
Since #13666 has been merged, I get back to this PR.

I rebased this PR on master and included a ztimer implementation for `evtimer_now_*()` .

Furthermore, I've tested `tests/evtimer_underflow` on the `samr30-xpro` board and it turns out to be working properly. I guess ztimer isn't configured for `native` properly, yet. I'll further investigate this issue.

File: tests/evtimer_msg/Makefile
Comment (maribu) [inline tests/evtimer_msg/Makefile:4]: This should not be needed. `xtimer` is (for now) still the default high level timer API an should be pulled in as dependency, unless the user explicitly wishes to do differently.

File: sys/ztimer/Makefile.dep
Comment (maribu) [inline sys/ztimer/Makefile.dep:44]: This should be moved to `makefiles/pseudomodules.mk`

File: sys/ztimer/Makefile.dep
Comment (maribu) [inline sys/ztimer/Makefile.dep:46]: This should be moved next to where the other `evtimer` dependencies are handled.

File: Makefile.dep
Comment (maribu) [inline Makefile.dep:692]: Please move the other dependency handling (see below) here, so that all `evtimer` related dependency management comes directly below `ifneq (,$(filter evtimer,$(USEMODULE)))`

File: sys/ztimer/Makefile.dep
Comment (maribu) [inline sys/ztimer/Makefile.dep:30]: Can you split this out into a separate commit, as this fix of indent is unrelated? Keeping unrelated stuff into separate commits makes life much easier when after merging some bug pops up and specific commits need to be reverted until a long term solution is found.

(Actually, unrelated stuff should be moved into its own PR. But I personally have strong sympathy for not wanting to create a PR for trivial single line fixes.)

**Comment by maribu** [comment]
Btw: It is now called `ztimer_xtimer_compat`, not `xtimer_on_ztimer`. Maybe it would make sense to rename this for consistency?

**Comment by kaspar030** [comment]
> Btw: It is now called `ztimer_xtimer_compat`, not `xtimer_on_ztimer`. Maybe it would make sense to rename this for consistency?

they both exist. "xtimer_on_ztimer" includes full xtimer, but makes xtimer use ztimer as backend. "ztimer_xtimer_compat" is a wrapper (xtimer calls get wrapped to ztimer calls). The latter doesn't support the complete xtimer api.

File: sys/ztimer/Makefile.dep
Comment (jue89) [reply sys/ztimer/Makefile.dep:44]: Ok. Should I also move the `xtimer_on_ztimer` pseudo module to that file?
Otherwise this isn't consistent ;)

File: tests/evtimer_msg/Makefile
Comment (jue89) [reply tests/evtimer_msg/Makefile:4]: Yes, right. But the test application makes direct use of `xtimer` methods. Thus, it has to be pulled in explicitly. Otherwise this test cannot be built when `evtimer` uses `ztimer_msec` as source.

File: sys/ztimer/Makefile.dep
Comment (jue89) [reply sys/ztimer/Makefile.dep:30]: Done

File: sys/ztimer/Makefile.dep
Comment (jue89) [reply sys/ztimer/Makefile.dep:46]: Same question: Also move the `xtimer_on_ztimer` dependency handling to the place where `xtimer` dependencies get handled?

File: sys/ztimer/Makefile.dep
Comment (maribu) [reply sys/ztimer/Makefile.dep:44]: Yes, this should certainly be fixed. It is currently the only place in RIOT `PSEUDOMODULES` is touched within a `Makefile.dep` and I really would avoid this pattern from sneaking in. The problem with those `Makefile.dep`s is that they are recursively included again and again until the content in `USEMODULE` (after sorting and de-duplicating) converges. And e.g. for `make info-boards-supported` the dependency resolution is done for each board RIOT supports. Thus, the `Makefile.dep`s should be as slim as possible.

But I would like to see this in a separate PR. No need to delay this PR for that. Do you want to create this PR? (If so, you could cherry-pick the indent fix also into that. It would be nice to also extend `dist/tools/buildsystem_sanity_check/check.sh` to enforce that the pattern `PSEUDOMODULE[\t ]*[+:]*=` is never found in any `Makefile.dep`. I would personally also would like to relocate those in the `Makefile.include` for consistency, but that is not so pressing as those in the `Makefile.dep`.)

Comment (maribu) [reply sys/ztimer/Makefile.dep:44]: (Please also state if you prefer not to open the PR. I would create the PR in that case.)

File: sys/ztimer/Makefile.dep
Comment (maribu) [reply sys/ztimer/Makefile.dep:46]: Hmm. I'm not so sure anymore. There is also some logic to having the `ztimer` related stuff (even if only indirect) here.

But I agree that either both `xtimer_on_ztimer` and `evtimer_on_ztimer` should be here, or both should be moved to `xtimer` or `evtimer`, respectively.

Comment (jue89) [reply sys/ztimer/Makefile.dep:46]: After I slept over it, I would prefer to have the whole dependency resolution for `xtimer` resp. `evtimer` at one place. Currently, for `xtimer` one half can be found in `Makefile.dep` and the other half at `sys/ztimer/Makefile.dep`. (Maybe @kaspar030 has a good reason for that?)

I think the best way is to open another PR fixing this for `xtimer`. Do you agree?

Another question, that just came into my mind: Why is most of the `sys/*` dependency resolution handled in the `Makefile.dep` and not in the `sys/Makefile.dep`? Is there a reason for that? I find this a little bit confusing.

File: sys/ztimer/Makefile.dep
Comment (jue89) [reply sys/ztimer/Makefile.dep:44]: Thanks for the detailed explanation!
I'll open a separate PR for that.

File: sys/ztimer/Makefile.dep
Comment (maribu) [reply sys/ztimer/Makefile.dep:46]: > Is there a reason for that?

My guess is: In beginning nobody saw a reason to split it up until it finally became a mess. Now that it became obvious that this file should be split out, new PRs add stuff to `sys/Makefile.dep`. But nobody did yet clean up the old stuff.

File: sys/ztimer/Makefile.dep
Comment (jue89) [reply sys/ztimer/Makefile.dep]: Done

**Comment by jue89** [comment]
I think everything that should be in this PR has been implemented, @maribu.

This PR:
 * Implements the `evtimer_on_ztimer`
 * Handels the `evtimer_on_ztimer` dependency management in `sys/ztimer/Makefile.dep`
 * Fixes `tests/evtimer_msg` when `evtimer`

Follow-up PR 1:
 * Move `xtimer_on_ztimer` pseudo module definition to `makefiles/pseudomodules.mk`
 * Implement a check that no `Makefile.dep` touches `PSEUDOMODULES`

Follow-up PR 2:
 * Move `xtimer` and `xtimer_on_ztimer` dependency management to `sys/Makefile.dep`
 * Move `evtimer` and `evtimer_on_ztimer` dependency management to `sys/Makefile.dep`
 * Fix the indentation quirk

Does this sound like a plan? If so I would rebase and clean-up this PR if you don't mind. It became kinda messy ;)

**Comment by kaspar030** [comment]
>     * Implement a check that no `Makefile.dep` touches `PSEUDOMODULES`

We could also go the other way around. There's no harm in Makefile.dep adding to pseudomodules. We could move them from pseudomodules.mk to the Makefile.deps, so they're close to the other module dependencies.

**Comment by jue89** [comment]
Rebased and ready for review ;)

**Comment by jue89** [comment]
> We could also go the other way around. There's no harm in Makefile.dep adding to pseudomodules. We could move them from pseudomodules.mk to the Makefile.deps, so they're close to the other module dependencies.

I think @maribu's point is the recursive dependency resolution by including all `Makefile.dep` over and over again until the result hasn't altered during a run. So in this case `PSEUDOMODULES` ends up with containing `xtimer_on_ztimer` several times.

This does work but takes some time. But in case of the conditional `PSEUDOMODULES += xtimer` I don't see any other solution to pull `PSEUDOMODULES` into the dependency resolution process and deduplicate und sort it, as well.

Or have I got something wrong?

**Comment by fjmolinas** [comment]
> This does work but takes some time. But in case of the conditional PSEUDOMODULES += xtimer I don't see any other solution to pull PSEUDOMODULES into the dependency resolution process and deduplicate und sort it, as well.

Could it work if you declare a pseudomodule `xtimer_on_xtimer` and then in `xtimer/Makefile`?

```
NO_AUTO_SRC = 1

ifneq (,$(filter xtimer_on_xtimer,$(USEMODULE)))
  SRC += xtimer_core.c 
  SRC += xtimer.c 
endif

include $(RIOTBASE)/Makefile.base
```

(did not test this, just brainstorming)

**Comment by maribu** [comment]
> But in case of the conditional `PSEUDOMODULES += xtimer` I don't see any other solution to pull `PSEUDOMODULES` into the dependency resolution process and deduplicate und sort it, as well.

Oh yes, a module this is only conditionally a pseudo-module is indeed a special case. The approach @fjmolinas suggested would certainly work. Maybe it could be simplified to this:

```Makefile
ifneq (,$(filter xtimer_on_xtimer,$(USEMODULE)))
  NO_AUTO_SRC = 1
endif
```

Alternatively, this special case IMO justifies an exception. A small comment above the `PSEUDOMODULES +=` outside of `pseudomodules.mk` to state the reasons for this would be greatly appreciated. This would prevent it for becoming a role model for touching `PSEUDOMODULES` during dependency resolution.

I don't think that de-duplication of `PSEUDOMODULES` would actually be needed. If I recall correctly, we only access it using `$(filter <FOO>,$(PSEUDOMODULES))`; and for that it does not matter if it contains duplicates. (Deduplication might improve performance if the array would contain many duplicates, though. But that is not the case here.)

**Comment by jue89** [comment]
Really good idea! But I would put this in a follow-up PR.
(And the name of the trouble maker is `ztimer_xtimer_compat` ;))

**Comment by kaspar030** [comment]
> I think @maribu's point is the recursive dependency resolution by including all `Makefile.dep` over and over again until the result hasn't altered during a run. So in this case `PSEUDOMODULES` ends up with containing `xtimer_on_ztimer` several times.
> 
> This does work but takes some time.

Can we quantify this? The recursion stops when there are no changes (additions) to the checked variables. Adding to PSEUDOMODULES doesn't even cause deeper recursion. And if having a module in there multiple times probably won't have any measurable effect on performance. Neither would a `PSEUDOMODULES := $(sort $ÜSEUDOMODULES))`, which would deduplicate it.

You can certainly "fix" this, but IMO there's no issue.

**Comment by maribu** [comment]
> Can we quantify this? The recursion stops when there are no changes (additions) to the checked variables.

We recently had a situation in which `make info-boards-supported` took over 15 seconds, without the number of dependency recursions being increased. With roughly 200 boards supported and assuming the recursion stops on average after 5 steps, something taking 1ms during dependency resolution will result in `make info-boards-supported` taking one second longer. When the CI runs `make info-boards-supported` for the 385 tests an additional millisecond spent in `Makefile.dep` becomes 6 minutes 25 seconds of additional CPU time. (Of course, the CI will run on many workers, but the impact should still be noticeable.)

IMO, we really should try hard to avoid bloating up the `Makefile.dep`s with stuff that can be evaluated only once, instead of thousand times per application.

**Comment by kaspar030** [comment]
> We recently had a situation in which `make info-boards-supported` took over 15 seconds, without the number of dependency recursions being increased. With roughly 200 boards supported and assuming the recursion stops on average after 5 steps, something taking 1ms during dependency resolution will result in `make info-boards-supported` taking one second longer. When the CI runs `make info-boards-supported` for the 385 tests an additional millisecond spent in `Makefile.dep` becomes 6 minutes 25 seconds of additional CPU time. (Of course, the CI will run on many workers, but the impact should still be noticeable.)

Yeah, well, adding to pseudomodules doesn't increase recursion depth. I'm just saying.

**Comment by jue89** [comment]
Hmm, I can understand both sides. And I tend to opt for the better maintainable solution than the faster one. It took me a few months working with RIOT until I found the `pseudomodules.mk` ;)

**Comment by maribu** [comment]
> It took me a few months working with RIOT until I found the `pseudomodules.mk` ;)

This might indicate an documentation issue.

In any case: Having most `PSEUDOMODULES` defined in `pseudomodules.mk` with some random exceptions is clearly not increasing maintainability, but confusion.

A PR to reorganize where `PSEUDOMODULES` are defined and documenting this properly is certainly welcome. But until a general solution has been agreed upon, let's ***please*** not mess things up even more by picking a random set of pseudo modules and declare them outside of `pseudomodules.mk`.

**Comment by fjmolinas** [comment]
> > It took me a few months working with RIOT until I found the `pseudomodules.mk` ;)
> 
> This might indicate an documentation issue.

The doc we have for pseudo-modules (that is not in code) is currently here http://riot-os.org/api/creating-modules.html#pseudomodules. Is it just incomplete or is it  not visible enough?

**Comment by kaspar030** [comment]
> (did not test this, just brainstorming)

`sys/xtimer` is not evaluated if ztimer_xtimer_compat is used, see https://github.com/RIOT-OS/RIOT/blob/9f707bf1212fef0686315a1623c5371a2aa8ef78/sys/Makefile#L79-L82.

**Comment by kaspar030** [comment]
> In any case: Having most `PSEUDOMODULES` defined in `pseudomodules.mk` with some random exceptions is clearly not increasing maintainability, but confusion.

I don't agree. complexity adds to maintanance burden and confusion. Adding to PSEUDOMODULES in ztimer's Makefile.dep is simple and fast. restricting adding to PSEUDOMODULES in pseudomodules.mk is nowadays an arbitrary, confusing restriction.

The one thing that PSEUDOMODULES does is removing "module_name.a" from the linking command line.

**Comment by jue89** [comment]
> > > It took me a few months working with RIOT until I found the `pseudomodules.mk` ;)
> > 
> > 
> > This might indicate an documentation issue.
> 
> The doc we have for pseudo-modules (that is not in code) is currently here http://riot-os.org/api/creating-modules.html#pseudomodules. Is it just incomplete or is it not visible enough?

To be honest: I haven't seen this bit, because I thought that I understood the modules system by looking into different Makefiles combined with some educated guess. Thus, the current solution is not intuitive to my eyes. I wasn't expecting dependency handling of one module at two completely different paths / folders.

@maribu @kaspar030 Wouldn't it be a quick experiment to move the entire content of `makefiles/pseudomodules.mk` into `Makefile.dep` and measure the additional required time?

**Comment by maribu** [comment]
> The one thing that PSEUDOMODULES does is removing "module_name.a" from the linking command line.

The semantics are irrelevant. The problem discussed is will it add confusion if definitions are not where people expect them to be or even where the documentation states them to search for them:

> To create a pseudomodule just add its name to makefiles/pseudomodules.inc.mk with PSEUDOMODULES += <modulename> in alphabetical order.

As [I said](https://github.com/RIOT-OS/RIOT/pull/13661#issuecomment-630787017), I have nothing against making an exception for the very specific case of a module being a pseudo module only on specific conditions. But we should document this with a comment.

If we in general want to restructure where pseudo modules live (and not just specific justified exceptions), we should discuss this independently of this PR and transition to the new structure consistently, rather than creating a huge mess with mixing different structures.

**Comment by fjmolinas** [comment]
@jue89 with #14123 maybe this one needs a rebase!

**Comment by jue89** [comment]
And rebased :)

File: sys/include/evtimer.h
Comment (kaspar030) [inline sys/include/evtimer.h:80]:
```suggestion
    uint32_t base;                  /**< Absolute time the first event is built on */
```

File: sys/include/evtimer.h
Comment (kaspar030) [inline sys/include/evtimer.h:138]: this is a bit dangerous, on ztimer, this will overflow after 49 days. with xtimer_now_usec64(), not in our lifetimes.
A quickfix would be to depend on ztimer_now64.

Comment (kaspar030) [reply sys/include/evtimer.h:138]: is there a user for this function? :)

**Comment by kaspar030** [comment]
Apart from the evtimer minutes thing, IMO this is good to go.

File: sys/include/evtimer.h
Comment (jue89) [reply sys/include/evtimer.h:138]: Good point! Done.
`gnrc_ipv6_nib` is a user of this method. ;)

**Comment by jue89** [comment]
I'll setup a little testbed with my `samr30-xpro` and `examples/gnrc_minimal` to test if this will survive for some days.

**Comment by jue89** [comment]
> I'll setup a little testbed with my `samr30-xpro` and `examples/gnrc_minimal` to test if this will survive for some days.

First results after the weekend: Looks good. The board still lives. I tested this by pinging its IP.

Setup:
* Set `ENABLE_DEBUG` to 1 in `sys/evtimer/evtimer.c`
* Set `USEMODULE="ztimer ztimer_usec evtimer_on_ztimer` during compilation
* Observe the UART interface for debug messages

**Comment by fjmolinas** [comment]
@kaspar030 only squashing is missing here, are you ok with the change?

**Comment by maribu** [comment]
Note: This won't change generated machine code, unless someone explicitly choose to use `ztimer`. IMO, this is, hence, low impact and should be considered for merging even during soft feature freeze.

**Comment by bergzand** [comment]
> Note: This won't change generated machine code, unless someone explicitly choose to use `ztimer`. IMO, this is, hence, low impact and should be considered for merging even during soft feature freeze.

Fine with me :+1:

**Comment by kaspar030** [comment]
yes, please squash!

**Comment by benpicco** [comment]
Looks like this uncovered a missing `#include <assert.h` in `gnrc/mac/timeout.h`

**Comment by jue89** [comment]
> Looks like this uncovered a missing `#include <assert.h` in `gnrc/mac/timeout.h`

Yip. I pushed a commit fixing that. I think this one-liner doesn't need its own PR?

**Comment by kaspar030** [comment]
Should we update the documentation in this PR or postpone to a followup? It is currently pretty specific with explaining that xtimer is used as backend...

**Comment by jue89** [comment]
Hmm, I'll push some lines the evtimer's documentation pointing out that it may be based on `xtimer` or `ztimer_msec` depending on configuration. Currently the documentation states "uses xtimer as backend" which isn't true for this case ;)

Postponing documentation to a later point in time is likely to never happen :D

**Comment by jue89** [comment]
I added a quick note to refer to `evtimer_on_ztimer`. Now, nothing wrong is stated in the documentation. Should I add more docu or is this already enough?

**Comment by maribu** [comment]
> I added a quick note to refer to `evtimer_on_ztimer`. Now, nothing wrong is stated in the documentation. Should I add more docu or is this already enough?

IMO, this is good as it as. Let's rerun Murdock. (One build failed due to connectivity issues of the worker.)

**Comment by maribu** [comment]
@kaspar030: All green ;-)

**Review by kaspar030 — APPROVED** [review]
ACK.

**Comment by jue89** [comment]
Am I allowed to press that green shiny Merge button? :)

**Comment by jue89** [comment]
Thank you all for reviewing!

File: sys/ztimer/Makefile.dep
Comment (kfessel) [inline sys/ztimer/Makefile.dep:46]: why was this line added?

Comment (kfessel) [reply sys/ztimer/Makefile.dep:46]: speaking about that  `ztimer_now64`

**Pull request merged**
